include/ginkgo/core/matrix/row_gatherer.hpp

```diff
@@ -0,0 +1,181 @@
+
+
+/**
+ * RowGatherer is a matrix "format" which stores the row rowgatherer
+ * arrays which can be used to perform rowgatherer rows to another matrix.
+ *
+ * @tparam IndexType  precision of rowgatherer array indices.
+ *
+ * @note This format is used mainly to allow for an abstraction of the
+ * rowgatherer and provides the user with an apply method which
+ * calls the respective LinOp's rowgatherer operation if the respective LinOp
+ * implements the RowGatherable interface. As such it only stores an array of
+ * the rowgatherer indices.
+ *
+ * @ingroup rowgatherer
+ * @ingroup matrix
+ * @ingroup LinOp
+ */
+template <typename IndexType = int32>
+class RowGatherer : public EnableLinOp<RowGatherer<IndexType>>,
+                    public EnableCreateMethod<RowGatherer<IndexType>> {
+    friend class EnableCreateMethod<RowGatherer>;
+    friend class EnablePolymorphicObject<RowGatherer, LinOp>;
+
+public:
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the array of row_gather_index.
+     *
+     * @return the pointer to the row_gather_index array.
+     */
+    index_type* get_row_gather_index() noexcept
+    {
+        return row_gather_index_.get_data();
+    }
+
+    /**
+     * @copydoc get_row_gather_index()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type* get_const_row_gather_index() const noexcept
+    {
+        return row_gather_index_.get_const_data();
+    }
+
+    /**
+     * Returns the number of elements explicitly stored in the row_gather_index
+     * array.
+     *
+     * @return the number of elements explicitly stored in the row_gather_index
+     * array.
+     */
+    size_type get_row_gather_index_size() const noexcept
+    {
+        return row_gather_index_.get_num_elems();
+    }
+
+protected:
+    /**
+     * Creates an uninitialized RowGatherer arrays on the specified executor..
+     *
+     * @param exec  Executor associated to the LinOp
+     */
+    RowGatherer(std::shared_ptr<const Executor> exec)
+        : RowGatherer(std::move(exec), dim<2>{})
+    {}
+
+    /**
+     * Creates uninitialized RowGatherer arrays of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the RowGatherable matrix
+     */
+    RowGatherer(std::shared_ptr<const Executor> exec, const dim<2>& size)
+        : EnableLinOp<RowGatherer>(exec, size), row_gather_index_(exec, size[0])
+    {}
+
+    /**
+     * Creates a RowGatherer matrix from an already allocated (and initialized)
+     * row and column rowgatherer arrays.
+     *
+     * @tparam IndicesArray  type of array of indices
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the rowgatherer array.
+     * @param row_gather_indexindices array of rowgatherer array
```

Comment:
If you want to stick with singular, then
```suggestion
     * @param row_gather_index array of rowgatherer array
```